service/src/test/java/bio/terra/cbas/controllers/util/TestRunSetHelper.java

```diff
@@ -0,0 +1,373 @@
+      new WdsRecordSet().recordType(recordType).recordIds(List.of(recordId1, recordId2));
+
+  private final RunSetRequest runSetRequest =
+      new RunSetRequest()
+          .runSetName("mock-run-set")
+          .methodVersionId(methodVersionId)
+          .workflowInputDefinitions(List.of(input1, input2))
+          .wdsRecords(wdsRecordSet);
+
+  private final Map<String, UUID> recordIdToRunIdMapping =
+      new HashMap<>() {
+        {
+          put(recordId1, runId1);
+          put(recordId2, runId2);
+        }
+      };
+
+  @Test
+  void submissionLaunchesSuccessfully() throws Exception {
+    // Set up WDS API responses
+    when(wdsService.getRecord(eq(recordType), eq(recordId1), any()))
+        .thenReturn(
+            new RecordResponse().type(recordType).id(recordId1).attributes(recordAttributes1));
+    when(wdsService.getRecord(eq(recordType), eq(recordId2), any()))
+        .thenReturn(
+            new RecordResponse().type(recordType).id(recordId2).attributes(recordAttributes2));
+
+    when(cromwellService.submitWorkflowBatch(eq(mockWorkflowUrl), any(), any(), any()))
+        .thenReturn(
+            List.of(
+                new WorkflowIdAndStatus().id(engineId1.toString()).status("Running"),
+                new WorkflowIdAndStatus().id(engineId2.toString()).status("Running")));
+
+    when(cbasApiConfiguration.getMaxWorkflowsInBatch()).thenReturn(10);
+    when(uuidSource.generateUUID()).thenReturn(engineId1).thenReturn(engineId2);
+
+    mockRunSetHelper.triggerWorkflowSubmission(
+        runSetRequest, runSet, recordIdToRunIdMapping, mockToken, mockWorkflowUrl);
+
+    // verify that Runs were set to Initializing state
+    verify(runDao)
+        .updateEngineIdAndRunStatus(
+            eq(runId1), eq(engineId1), eq(CbasRunStatus.INITIALIZING), any());
+    verify(runDao)
+        .updateEngineIdAndRunStatus(
+            eq(runId2), eq(engineId2), eq(CbasRunStatus.INITIALIZING), any());
+
+    // verify that RunSet was set to Running state
+    verify(runSetDao)
+        .updateStateAndRunSetDetails(
+            eq(runSetId), eq(CbasRunSetStatus.RUNNING), eq(2), eq(0), any());
+  }
+
+  @Test
+  void wdsErrorDuringSubmission() throws Exception {
+    // Set up WDS API responses
+    when(wdsService.getRecord(eq(recordType), eq(recordId1), any()))
+        .thenReturn(
+            new RecordResponse().type(recordType).id(recordId1).attributes(recordAttributes1));
+    // return error when fetching record data for record ID 2
+    when(wdsService.getRecord(eq(recordType), eq(recordId2), any()))
+        .thenThrow(
+            new WdsServiceApiException(
+                new org.databiosphere.workspacedata.client.ApiException(
+                    400, "ApiException thrown for testing purposes.")));
+
+    when(runDao.getRuns(new RunDao.RunsFilters(runSetId, List.of(QUEUED))))
+        .thenReturn(List.of(run1, run2));
+
+    mockRunSetHelper.triggerWorkflowSubmission(
```

Comment:
Presumably in the context of this test case, this is _not_ running async?

Reply:
Yes that is correct since this is just testing the method directly.